server/etcdserver/api/v3rpc/health.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3rpc
+
+import (
+	"go.uber.org/zap"
+	"google.golang.org/grpc/health"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+)
+
+const (
+	allGRPCServices = ""
+)
+
+type HealthNotifier interface {
```

Comment:
Idea for followup. Instead of allowing defrag to change StartServe status as they will, better would be to just notify that defrag is happening and let health notify decide on health status. https://en.wikipedia.org/wiki/Law_of_Demeter, why defrag should know how about changing health.

Reply:
Additional improvement would be that Maintenance server would no longer need to be aware of `stopServingOnDefrag` flag, instead configuration would be handled by health.

Reply:
Thanks! I like the idea of decoupling stop serving from the maintenance server. 

Eventually the health notifier may extend to other readyz checks (it need more in-depth thought though) implemented in the readyz handler.